### ISSUE TITLE
[ Rush ] Generate a change log for projects that don't publish via npm

### ISSUE THREAD

**Issue by light24bulbs** [Microsoft/web-build-tools#884]
We are using Rush to build a microservice cluster.  Some rush packages will deploy to NPM, but after that some others(which depend on the above packages) will be containerized and go to AWS.  I'm guessing this is a pretty normal pattern.

It would be really great to use Rush to check what microservices have changed and just deploy those(if they have dockerfiles). 

There might be some clever use of version policy to achieve this. Looking for your recommendation.

In Lerna I can do `lerna ls --since --json` to see what I need to redeploy.

**Comment by pgonzal** [comment]
Is the idea that certain packages are released by running `npm publish`, but certain other packages are released via some custom script such as `some-docker-tool deploy-this-folder`?

In other words you want to generalize `rush publish` to support other release types besides NPM?

@iclanton and are really wanting to do a sprint to improve the publishing/versioning features.  We could try to incorporate this requirement into the design.  We already use `npm pack` instead of `npm publish` in our own repos.

**Comment by light24bulbs** [comment]
Yes that sums it up very well.  Getting all the updated packages in a parseable list is probably enough.

In lerna our pattern was:
1. `lerna ls --since --json` and store the list of updated packages
2. `lerna publish` to take all the packages not marked as `private:true` to npm repository
3. Then we looped through all the changed packages from step one, checking for dockerfiles and then deploying them all in one parallel job.  

When I saw the rush change files, I thought "this is perfect".  But then I realized change files are only generated if shouldPublish is true. Otherwise the `change` generator never gets triggered and versions never bump. 

Could we do something tricky like have a version policy called `npm` and a policy called `container` and then publish them differently in two different `publish` commands?

**Comment by pgonzal** [comment]
What does `--since` do exactly? I couldn't find any documentation for it.

> Could we do something tricky like have a version policy called npm and a policy called container and then publish them differently in two different publish commands?

At the point where Rush calls `npm publish`,   we could just provide a way to customize what command it runs. It could be a version policy option (that affects all packages in the group) or we could provide project-specific options (if you really needed that).

We already encountered this problem ourselves. Someone added [_updateApiFile](https://github.com/Microsoft/web-build-tools/blob/master/apps/rush-lib/src/cli/actions/PublishAction.ts) into Rush one time when they needed custom logic to run before publishing, and it really doesn't belong there. Also the `_npmPack()` felt a little out of place to me as well.

**Comment by light24bulbs** [comment]
That's good to know it's a common need!  I figure a lot of people would like to use rush this way.

--since isn't super well documented.  Without an argument, it filters `list`/`ls` to just give the packages that have changed since their last tag.  https://www.npmjs.com/package/@lerna/filter-options  Pretty much perfect for figuring out what to deploy.

Running a custom command is probably good enough.  What we really need is a list of all the packages that have changed, because our deployment system can take a list of containers to deploy in parallel.  We could have the script command write `>>` its package name to a file and then read that or something.  

If you're thinking about how it SHOULD work ideally, my first thought was adding a `shouldDeploy` boolean into the rush project config, and then I could create changeFiles and ask `rush whatShouldIdeploy` but I'm new to rush so that was just my first thought.  

I will play with version policies a little more and see if I can do what's needed right of the box.

**Comment by pgonzal** [comment]
> --since isn't super well documented.  Without an argument, it filters `list`/`ls` to just give the packages that have changed since their last tag.  https://www.npmjs.com/package/@lerna/filter-options  Pretty much perfect for figuring out what to deploy.

Is that because deploying creates a Git tag?  Sorry if these are dumb questions -- I really should spend more time reading about Lerna.  :-)

> I will play with version policies a little more and see if I can do what's needed right of the box.

Version policies are a much better approach to this.  I apologize that their current design is fairly confusing (since the requirements were constantly evolving for the people who implemented that portion of Rush).  We really need to do another iteration to make it more understandable.

**Comment by light24bulbs** [comment]
I believe git tag is how it works by default. Deploying does create a git
tag for the whole repo, creating an easy way for Lerna to check for changes
since the last release(I think, haven't read all the source).

I'm glad you said that because I am getting pretty confused by version
policy.  It seems like it is one flag trying to do a lot.  I'm trying to
use it as something I would call 'publishing group'.  Maybe it could just
use a dedicated doc page that really explains it.

**Comment by pgonzal** [comment]
So the node_modules folder is not included in your docker container?  Instead you run `pnpm install` as a postprocess after you boot up the container?

With PNPM's model, I believe it would be fairly straightforward to extract a project-specific shrinkwrap.yaml file for a given project. @zkochan does your "shared shrinkwrap file" feature help with this at all?

**Comment by light24bulbs** [comment]
We just figured out we can use 
```bash
cp -RLf node_modules local_node_modules && rm -rf node_modules && mv local_node_modules node_modules
```
To convert rush's symlinks into real files before containerization.  It's great because it avoids another install, respects the global package lock, and will also containerize any local packages that are linked but not in NPM.  

It would be trivial for rush to someday have a command `rush isolate` that did this. :)

**Comment by pgonzal** [comment]
@jbcpollak (who was also interested in Docker container deployment) FYI

> It would be trivial for rush to someday have a command `rush isolate` that did this. :)

Today you could probably implement this as a Rush [custom command](https://rushjs.io/pages/maintainer/custom_commands/).

Eventually we do want to generalize Rush's publishing scenario to support other release types besides `npm publish`, and it seems like Docker containers should probably be first class citizens.

**Comment by zkochan** [comment]
> With PNPM's model, I believe it would be fairly straightforward to distill a project-specific shrinkwrap.yaml file for a given project. @zkochan does your "shared shrinkwrap file" feature help with this at all?

A couple of days ago @hulkish was asking about this feature for the same pupose in [our gitter chat](https://gitter.im/pnpm/pnpm?at=5bbafe21600c5f64236ec18a)

Indeed, it would be easy to implement a pnpm command for fetching a dedicated shrinkwrap from shared shrinkwrap. We have that logic already in the [pnpm-shrinkwrap package](https://github.com/pnpm/pnpm/blob/064c58613fac67241f9726a0da3b9173aee4c928/packages/shrinkwrap/src/filter.ts#L11)

**Comment by ledniy** [comment]
@light24bulbs Hi, I guess I have the same problem as you, did you solve the problem with a list of changes? I'm still confusing with version policy and don't understand how to use it properly for our needs.

**Comment by RichiCoder1** [comment]
Not to necro an old thread, but was there an answer for this? I'd like to publish https://aws.github.io/jsii/ projects. So I only want published changed workspaces, but when I publish I want to take over the publish process to [jsii-pacmak](https://aws.github.io/jsii/user-guides/lib-author/quick-start/write-code/) to actually create the packages and manually publish.

**Comment by mathiasmaerker** [comment]
I would love to have a publish hook as well, I have a monorepo with very different stuff going on, e.g. publish to npm, publish to docker, do some terraform IaC etc. would be nice to have this feature

**Comment by witcher112** [comment]
I just came across this issue, are there any news / new solutions that resolve the problem described here?
